src/m365/tenant/commands/auditlog/auditlog-report.ts

```diff
@@ -0,0 +1,246 @@
+        if (!hasActiveSubscription) {
+          if (this.verbose) {
+            logger.logToStderr(`Starting subscription since subscription is not active for the content type`);
+          }
+          let startSubscriptionEndPoint: string = `activity/feed/subscriptions/start?contentType=${(<any>AuditContentTypes)[args.options.contentType]}&PublisherIdentifier=${this.tenantId}`;
+          const requestOptions: any = {
+            url: `${this.serviceUrl}/${this.tenantId}/${startSubscriptionEndPoint}`,
+            headers: {
+              accept: 'application/json;odata.metadata=none'
+            },
+            responseType: 'json'
+          };
+
+          return request.post(requestOptions);
+        }
+
+        return Promise.resolve();
+      });
+  }
+
+  private getAuditContentList(args: CommandArgs, logger: Logger): Promise<AuditContentList[]> {
+    if (this.verbose) {
+      logger.logToStderr(`Start listing Audit Content URL`);
+    }
+
+    let SubscriptionListEndpoint: string = (typeof args.options.startTime !== 'undefined' && typeof args.options.endTime !== 'undefined') ?
+      `/activity/feed/subscriptions/content?contentType=${(<any>AuditContentTypes)[args.options.contentType]}&PublisherIdentifier=${this.tenantId}&starttime=${args.options.startTime}&endTime=${args.options.endTime}` :
+      `/activity/feed/subscriptions/content?contentType=${(<any>AuditContentTypes)[args.options.contentType]}&PublisherIdentifier=${this.tenantId}`;
+    const requestOptions: any = {
+      url: `${this.serviceUrl}/${this.tenantId}/${SubscriptionListEndpoint}`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    return request.get<AuditContentList[]>(requestOptions)
+  }
+
+  private async getBatchedAuditLogReports(logger: Logger, AuditContentLists: AuditContentList[]) : Promise<any> {
+    if (this.verbose) {
+      logger.logToStderr(`Start generating Audit Reports in batchwise manner`);
+    }
+    
+    //Batching Approach - Given Batch size is 10
+    const batchSize: number = 10;
+    let runningBatch: number = 0;
+    var CompleteAuditReportsUnflattened : any = [];
+    //Temporary - Will be removed in production
+    let tempMaximumCount: number = 20;
+
+    for (let i = 0; i < (AuditContentLists.length<tempMaximumCount?AuditContentLists.length:tempMaximumCount); i += batchSize) {
+      if (this.verbose) {
+        runningBatch++;
+        logger.logToStderr(`Generating Audit Reports for Batch : ${runningBatch}`);
+      }
+
+      const PromiseRequestBatch = AuditContentLists.slice(i, i + batchSize<AuditContentLists.length?i+batchSize:AuditContentLists.length).map((AuditContentList) => {
+        //logger.log(`Inner Loop : ${i}`);
+        return this.getAuditLogReportforSingleContentURL(AuditContentList.contentUri);
+      })
+      
+      //Async/AWAIT Approach
+      CompleteAuditReportsUnflattened.push(await Promise.all(PromiseRequestBatch));
```

Comment:
Just to give you more details on the implementaiton.
Here as per your [suggestion](https://github.com/pnp/cli-microsoft365/issues/1739#issuecomment-733880218), I have batched the multiple promises. 

- Typically we will get around 200 Records which will have different content URL. We have to query each content URL which will fetch you the result That is how the Activity Feed API works.

- When we query each URL, it may fetch around 165 odd records (Just a ballpark count)

- Here, we have batched these promises on the count of 10 which means,
- In our case there will be 20 batches.
- I had tried to give chained promises within a loop, but somehow I am unable to get in the right direction. 
- It will be great if you can give your valuable insights on how can we go ahead with that.

As of now  I tried giving ASYNC / AWAIT approach which is fetching the result. But since our test cases will not support that, we may have to go with alternatives. Can you give any suggestion on that please.